### ISSUE TITLE
Please update firmware to v0257 to support multi-drop

### ISSUE THREAD

**Issue by CoelacanthusHex** [wuxx/nanoDAP-HS#9]
Openocd added multi-drop support, but it [needs CMSIS-DAP support SWD_Sequeue](https://review.openocd.org/c/openocd/+/6117). It means we need a firmware that is newer than [this commit](https://github.com/ARMmbed/DAPLink/commit/e6a327efeb3005d981cbb2c44674d808752ce989), for version, it is v0257.
If no newer firmware, nanoDAP-HS can't debug hardware with multi-drop, such as Raspberry Pi Pico.
Hope nanoDAP-HS can be better and better.

**Comment by flaviut** [comment]
[According to the readme,](https://github.com/wuxx/nanoDAP-HS/blob/master/README_en.md#firmware-upgrade) the firmware is just DAPLink. You can compile it [using these instructions](https://github.com/ARMmbed/DAPLink/blob/main/docs/DEVELOPERS-GUIDE.md). The command is `python tools/progen_compile.py -v sam3u2c_if`.

I've been unable to test myself due to some issues with drag-and-drop programming on my device, but please tell me if it works.

**Comment by CoelacanthusHex** [comment]
> I've been unable to test myself due to some issues with drag-and-drop programming on my device, but please tell me if it works.

I tried to compile it by myself before but met the same problem.

**Comment by flaviut** [comment]
I spoke with wuxx on AliExpress, and I think I have the situation cleared up, at least for me. There's two debug probes models here, and the branding is confusing:

- The nanoDAP is about $6, uses an GD32F150G8 MCU, and doesn't support USB Mass Storage-based flashing. wuxx said that it runs proprietary firmware, although it looks like this is it? https://github.com/wuxx/nanoDAP/tree/master/software
- The nanoDAP-HS is about $15, uses an ATMEL SAM3U MCU, and supports USB Mass Storage-based flashing. I assume the upstream DAPLink works on this device.

I've also done a bit of my own poking around, and I think I have a pinout for the nanoDAP debug header:

![Mini-DAPLink-Debug-Probe-TYPE-C-STM32-NRF51-52-ARM-Cortex-M-MCU-JTAG-SWD-CDC jpg_Q90](https://user-images.githubusercontent.com/5213469/196036914-1b893a16-ea83-44bb-a91a-a945aaed82ba.jpg)

**Comment by CoelacanthusHex** [comment]
> I spoke with wuxx on AliExpress, and I think I have the situation cleared up, at least for me. There's two debug probes models here, and the branding is confusing:
> 
>     * The nanoDAP is about $6, uses an GD32F150G8 MCU, and doesn't support USB Mass Storage-based flashing. wuxx said that it runs proprietary firmware, although it looks like this is it? [wuxx/nanoDAP@`master`/software](https://github.com/wuxx/nanoDAP/tree/master/software?rgh-link-date=2022-10-16T13%3A05%3A16Z)
> 
>     * The nanoDAP-HS is about $15, uses an ATMEL SAM3U MCU, and supports USB Mass Storage-based flashing. I assume the upstream DAPLink works on this device.
> 
> 
> I've also done a bit of my own poking around, and I think I have a pinout for the nanoDAP debug header:
> 
> ![Mini-DAPLink-Debug-Probe-TYPE-C-STM32-NRF51-52-ARM-Cortex-M-MCU-JTAG-SWD-CDC jpg_Q90](https://user-images.githubusercontent.com/5213469/196036914-1b893a16-ea83-44bb-a91a-a945aaed82ba.jpg)

What I have is the nanoDAP-HS version, the board silkscreen is "nanoDAP-hs v1.3", and the SoC silkscreen is "Atmel ATSAM3U2C AU", but I still cannot use USB Mass Storage-based flashing with the official DAPLink firmware.

**Comment by ckrenslehner** [comment]
@flaviut did you actually manage to flash the rp2040 using the nanoDAP? 
I plan to update the firmware of the nanoDAP as you described above. :-)

I have version V1.5 but I always get a SWD Sequence expected error when I try to flash the rp2040 with pyocd..

**Comment by flaviut** [comment]
Yup, I've had great luck flashing my rp2040 with the nanodap-hs. Iirc I had to pull the latest upstream changes into the port's repo.

**Comment by BobbyCounts** [comment]
@flaviut Is the pin configuration different from what is default in daplink for the sam3u2c target? I tried compiling the latest daplink but when I upload I get nothing.

**Comment by flaviut** [comment]
I wish I could answer... I went through my files and I wasn't able to find the configuration I used. Sorry :(

**Comment by flaviut** [comment]
I recently needed to update a new device. First I checked out the latest tagged version: `git checkout v0257`, then [applied this patch](https://github.com/ARMmbed/DAPLink/issues/1003#issuecomment-1412320300):

```
diff --git a/records/tools/gcc_arm.yaml b/records/tools/gcc_arm.yaml
index 3ca532e6..0e54800d 100644
--- a/records/tools/gcc_arm.yaml
+++ b/records/tools/gcc_arm.yaml
@@ -31,6 +31,7 @@ tool_specific:
                 - -Wl,--gc-sections
                 - -Wl,--no-wchar-size-warning
                 - -Wl,--print-memory-usage
+                - -Wl,--no-warn-rwx-segment
             pre_build_script:
                 - tools/pre_build_script.py
             post_build_script:
diff --git a/source/daplink/sdk_stub.c b/source/daplink/sdk_stub.c
index e9116a63..87a2736b 100644
--- a/source/daplink/sdk_stub.c
+++ b/source/daplink/sdk_stub.c
@@ -25,3 +25,8 @@ __WEAK void sdk_init()
 {
     // Do nothing
 }
+
+void _close_r(void) {}
+void _lseek_r(void) {}
+void _read_r(void) {}
+void _write_r(void) {}
```

here's my bin file, [which I dragged into the MAINTENANCE drive](https://github.com/wuxx/nanoDAP-HS/blob/master/README_en.md#firmware-upgrade): [sam3u2c_if.bin.zip](https://github.com/wuxx/nanoDAP-HS/files/14968129/sam3u2c_if.bin.zip)

**Comment by hitsmaxft** [comment]
my nanodap-hs has a sam3u4c mcu on it , which isn't supported  by daplink. daplink officially supports sam3u2c
is here any port repo of daplink for this hardware?
